core/modules/drr.cc

```diff
@@ -0,0 +1,392 @@
+  pb_cmd_response_t response;
+  set_cmd_response_error(&response, SetQuantumSize(arg.quantum()));
+  return response;
+}
+
+pb_cmd_response_t DRR::CommandMaxFlowQueueSize(
+      const bess::pb::DRRMaxFlowQueueSizeArg &arg) {
+  pb_cmd_response_t response;
+  set_cmd_response_error(&response, SetMaxFlowQueueSize(arg.max_queue_size()));
+  return response;
+}
+
+void DRR::ProcessBatch(bess::PacketBatch *batch) {
+  int err = 0;
+
+  // insert packets in the batch into their corresponding flows
+  for (int i = 0; i < batch->cnt(); i++) {
+    bess::Packet *pkt = batch->pkts()[i];
+
+    // TODO(joshua): Add support for fragmented packets.
+    FlowId id = GetId(pkt);
+    auto it = flows_.Find(id);
+
+    // if the Flow doesn't exist create one
+    // and add the packet to the new Flow
+    if (it == nullptr) {
+      if (llring_full(flow_ring_)) {
+        bess::Packet::Free(pkt);
+      } else {
+        AddNewFlow(pkt, id, &err);
+        assert(err == 0);
+      }
+    } else {
+      Enqueue(it->second, pkt, &err);
+      assert(err == 0);
+    }
+  }
+}
+
+struct task_result DRR::RunTask(void *) {
+  bess::PacketBatch batch;
+  struct task_result ret;
+  int err = 0;
+  batch.clear();
+  int total_bytes = 0;
+  if (flow_ring_ != NULL) {
+     total_bytes = GetNextBatch(&batch, &err);
+  }
+  assert(err >= 0);  // TODO(joshua) do proper error checking
+
+  if (total_bytes > 0) {
+    RunNextModule(&batch);
+  }
+  
+  // the number of bits inserted into the packet batch
+  uint64_t cnt = batch.cnt();
+  uint64_t bits_retrieved = (total_bytes + cnt* kPacketOverhead) * 8;
+  ret = (struct task_result){
+      .packets = cnt, .bits = bits_retrieved 
+  };
+
+  return ret;
+}
+
+int DRR::GetNextBatch(bess::PacketBatch* batch, int* err) {
+  Flow* f;
+  uint32_t total_bytes = 0;
+  int count = llring_count(flow_ring_);
+  if(current_flow_) {
+    count++;
+  }
+  int batch_size = batch->cnt();
+
+  // iterate through flows in round robin fashion until batch is full
+  while (!batch->full()) {
+    // checks to see if there has been no update after a full round
+	// ensures that if every flow is empty or if there are no flows 
+	// that will terminate with a non-full batch.  
+    if (count == 0) {
+      if (batch_size == batch->cnt()) {
+        break;
+      } else {
+        count = llring_count(flow_ring_);
+        batch_size = batch->cnt();
+      }
+    }
+    count--;
+
+    f = GetNextFlow(err);
+    if(*err != 0) {
+      return total_bytes;
+    } else if(f == nullptr) {
+      continue;
+    }
+
+    int bytes = GetNextPackets(batch, f, err);
+    total_bytes += bytes;
+    if (*err != 0) {
+      return total_bytes;
+    }
+
+    if (llring_empty(f->queue) && !f->next_packet) {
+      f->deficit = 0;
+    }
+
+    // if the flow doesn't have any more packets to give, reenqueue it
+    if (!f->next_packet || f->next_packet->total_len() > f->deficit) {
+      *err = llring_sp_enqueue(flow_ring_, f);
```

Comment:
nit: Is there any reason you're using llring_s{p,c}_{en,de}queue instead of llring_{en,de}queue throughout the code? llring_enqueue()/llring_dequeue() both check how the ring was initialized and pick the correct enqueue/dequeue function. If it doesn't impose too much overhead I'd suggest using the latter instead. The code is slightly easier to maintain this way (though I doubt we'd ever switch to mp or mc).